include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+# include <openssl/safestack.h>
+# include "crypto/sparse_array.h"
+
+/*-
+ * Types and structures
+ * --------------------
+ */
+
+/*
+ * An event is expected to be very short lived, and may be a static
+ * (re-usable) variable, local in a function or other structure, or
+ * dynamically allocated.
+ */
+struct ossl_event_st;
+typedef int ossl_event_destructor_fn(struct ossl_event_st *event);
+
+/*
+ * Generic callback function type.  This is used for subscribers and filters
+ * alike, as seen in the different functions where this type is used.
+ */
+typedef int ossl_event_callback_fn(struct ossl_event_st *event,
+                                   void *callback_data);
+
+struct ossl_event_st {
+    /*
+     * Mandatory data.
+     *
+     * These fields are considered immutable.  However, the contents that
+     * |ctx| points at may be mutable.
+     */
+
+    /* A numeric event identity, defined by the using functionality */
+    uint32_t type;
+    /* A context relevant for the using functionality */
+    void *ctx;
+
+    /*
+     * Optional data (may be NULL).
+     *
+     * These fields are considered mutable, in so far that an event handler
+     * may "steal" them, and replace them with NULL.
+     */
+
+    /* Identifying material */
+    const void *identifiers;
+    /* Data carried by the event */
+    size_t payload_size;
+    void *payload;
+
+    /*
+     * A destructor for |identifiers| and |payload|, if necessary.
+     *
+     * This field is considered immutable until the event is destroyed.
+     */
+    ossl_event_destructor_fn *destructor;
+
+    /*
+     * Flags, only relevant for the event functionality itself.
+     */
+    unsigned int flag_dynamic : 1;
+};
+
+/*
+ * The event queue is expected to be very long lived, even thought it may be
+ * dynamically allocated.  Subsribers are expected to return values that
+ * express if the event it received was used or not.
+ */
+enum ossl_event_subscriber_returns {
+    /* The subscriber encountered an error */
+    ossl_event_error = -1,
+    /* The subscriber didn't use this event */
+    ossl_event_pass = 0,
+    /* The subscriber used this event */
+    ossl_event_handled = 1,
+};
+
+/*
+ * OSSL_EVENT_SUBSCRIBER_CLOSURE implements a closure simile, speciflcally
+ * for our event callbacks, to carry along the callback specific data pointer.
+ */
+typedef struct ossl_event_subscriber_closure_st {
+    ossl_event_callback_fn *cb;
+    void *cbarg;
+} OSSL_EVENT_SUBSCRIBER_CLOSURE;
+
+/*
+ * Opaque definition, to allow an optimal internal implementation without
+ * having to change anything else.
+ * This structure is expected to have the semantics of an array of stacks
+ * of events, and should only be affected by event addition, event removal
+ * and event dispatching, all of is internal code in crypto/event/event.c.
+ *
+ * Each element in this implied array represent a priority level, and each
+ * stack (element in this array) simply stores the events for its priority
```

Comment:
Noting that our safestack is inappropriate for the kinds of operations being done here.  A linked list or double ended queue would be far better.

FIFO on our stack is O(n).  Delete from the middle (unregistering an event handler) is also O(n).  Insert at back is O(1).

A linked list or double ended queue has O(1) for all of these.

Reply:
Good idea.  And that's exactly the reason I made this bit very opaque, so that the exact mechanism would be an implementation detail, with no concern here.  If the stack is implemented as a double linked list, it still has the essential behaviors of a stack: push and shift.

Reply:
FIFO implies queue semantics not stack.  Let's not mix up the terminology.